src/MongoDB.Driver.Core/Core/ChannelPinningHelper.cs

```diff
@@ -0,0 +1,174 @@
+        public static IReadBindingHandle CreateEffectiveReadBinding(ICluster cluster, ICoreSessionHandle session, ReadPreference readPreference)
+        {
+            IReadBinding readBinding;
+            if (session.IsInTransaction &&
+                IsConnectionPinned(session.CurrentTransaction) &&
+                session.CurrentTransaction.State != CoreTransactionState.Starting)
+            {
+                readBinding = new ChannelReadWriteBinding(
+                    session.CurrentTransaction.PinnedServer,
+                    session.CurrentTransaction.PinnedChannel,
+                    session);
+            }
+            else
+            {
+                if (IsInLoadBalancedMode(cluster.Description) && IsConnectionPinned(session.CurrentTransaction))
+                {
+                    // unpin if the next operation is not under transaction
+                    session.CurrentTransaction.UnpinAll();
+                }
+                readBinding = new ReadPreferenceBinding(cluster, readPreference, session);
+            }
+
+            return new ReadBindingHandle(readBinding);
+        }
+
+        /// <summary>
+        /// Create effective readwrite binding handle.
+        /// </summary>
+        /// <param name="cluster">The cluster.</param>
+        /// <param name="session">The session.</param>
+        /// <returns></returns>
+        public static IReadWriteBindingHandle CreateEffectiveReadWriteBinding(ICluster cluster, ICoreSessionHandle session)
+        {
+            IReadWriteBinding readWriteBinding;
+            if (session.IsInTransaction &&
+                IsConnectionPinned(session.CurrentTransaction) &&
+                session.CurrentTransaction.State != CoreTransactionState.Starting)
+            {
+                readWriteBinding = new ChannelReadWriteBinding(
+                    session.CurrentTransaction.PinnedServer,
+                    session.CurrentTransaction.PinnedChannel,
+                    session);
+            }
+            else
+            {
+                if (IsInLoadBalancedMode(cluster.Description) && IsConnectionPinned(session.CurrentTransaction))
+                {
+                    // unpin if the next operation is not under transaction
+                    session.CurrentTransaction.UnpinAll();
+                }
+                readWriteBinding = new WritableServerBinding(cluster, session);
+            }
+
+            return new ReadWriteBindingHandle(readWriteBinding);
+        }
+
+        internal static IChannelSourceHandle CreateEffectiveGetMoreChannelSource(IChannelSourceHandle channelSource, long cursorId)
+        {
+            IChannelSource effectiveChannelSource;
+            if (IsInLoadBalancedMode(channelSource.ServerDescription) && cursorId != 0)
+            {
+                var getMoreChannel = channelSource.GetChannel(CancellationToken.None); // no need for cancellation token since we already have channel in the source
```

Comment:
`channelSource` is configured here in line 127, so effectively we just call Fork for the forkedChannel variable in line 125